### PR TITLE
feat(scrapeConfigs): Add sharding to scrapeConfigs

### LINKED ISSUE #5637: ScrapeConfig objects don't work properly when Prometheus sharding is used
`generateScrapeConfig` gets the shard number as an input parameter but it doesn't make use of it. As a result, targets from ScrapeConfigs are scraped by all shards while they should be distributed.
 
https://github.com/prometheus-operator/prometheus-operator/blob/d3c37ea81bf3afcd7030819b054f65780e69a237/pkg/prometheus/promcfg.go#L2135-L2140

### PR DESCRIPTION
**Pull request by adinhodovic**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Fixes #5637.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
feat(scrapeConfigs): Add sharding to scrapeConfigs
```

### CODE REVIEW

**Review by simonpasquier — COMMENTED** [review]
Since ScrapeConfigs offer a low-level interface to the Prometheus configuration, I'd expect that users want to override the automated sharding in some cases. Which means that we may need something similar to what you implemented in 5c8245927a which checks the presence of hashmod action. WDYT?

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 
 	if sc.Spec.RelabelConfigs != nil {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if shards != 1 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2324]: I wouldn't make a special case for shard == 1

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+		alreadyHasShardLabel := false
+		if sc.Spec.RelabelConfigs != nil {
+			for _, relabelConfig := range sc.Spec.RelabelConfigs {
+				if relabelConfig.Action == "hashmod" {
+					level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "relabeling", relabelConfig)
+					alreadyHasShardLabel = true
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2330]:
```suggestion
					alreadyHasShardLabel = true
					break
```

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 
 	if sc.Spec.RelabelConfigs != nil {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if shards != 1 {
+		alreadyHasShardLabel := false
+		if sc.Spec.RelabelConfigs != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2326]: range works with nil slices too.
```suggestion
```

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 
 	if sc.Spec.RelabelConfigs != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2320]: (nit) not part of this change but checking the length of the slice would be more idiomatic.
```suggestion
	if len(sc.Spec.RelabelConfigs) > 0 {
```

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+					alreadyHasShardLabel = true
+				}
+			}
+		}
+		if !alreadyHasShardLabel {
+			relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
+		}
+	}
+
+	if sc.Spec.RelabelConfigs != nil || shards != 1 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2339]: do we need a condition here? At this point, `relabelings` has at least 1 item.

**Review by simonpasquier — COMMENTED** [review]
Thanks! Just a few nits.

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -2317,6 +2319,24 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 
 	if sc.Spec.RelabelConfigs != nil {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if shards != 1 {
+		alreadyHasShardLabel := false
+		if sc.Spec.RelabelConfigs != nil {
+			for _, relabelConfig := range sc.Spec.RelabelConfigs {
+				if relabelConfig.Action == "hashmod" {
```
Comment (leviliangtw) [inline pkg/prometheus/promcfg.go:2328]: Hello, I have a suggestion to make the code more readable. 
Here you check the action of "hashmod" as the indicator of "has Shard Label or not"
Is it possible to change the target label "__tmp_hash"  to something like "__tmp_shard_hash"?
Then, you can check the target label "__tmp_shard_hash" and make sure that this is of the purpose for sharding. 
Also, it could also avoid misjudgment if somebody just add their own "hashmod" action for other purpose.

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2328]: @leviliangtw I'm not sure to understand your point about the "__tmp_shard_hash" label?

I'm not aware of other use cases for the "hashmod" action than target sharding so I'm fine with the current approach but we might want to convey the nuance in the docs (e.g. "if the relabelConfigs list contains an item with `action: hashmod` then the operator will not apply the automatic target sharding.").

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (adinhodovic) [reply pkg/prometheus/promcfg.go:2324]: Had a similar approach to https://github.com/adinhodovic/prometheus-operator/blob/a65089818bb12c1d573650554b74d9fcc11e9fb5/pkg/prometheus/promcfg.go#L1668-L1670.

Fine with removing it, however the changeset is noisy due to all sharding relabel configs added to each scrape config as you can see in the commit.

File: pkg/prometheus/promcfg.go (hunk 5 above)
Comment (adinhodovic) [reply pkg/prometheus/promcfg.go]: No, not if we add sharding relabels for non-sharded setups. Fixed.

File: pkg/prometheus/promcfg.go [hunk 7]
```diff
@@ -2318,11 +2320,23 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+	for _, relabelConfig := range sc.Spec.RelabelConfigs {
+		if relabelConfig.Action == "hashmod" {
+			level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "relabeling", relabelConfig)
+			alreadyHasShardLabel = true
+			break
+		}
+	}
+	if !alreadyHasShardLabel {
+		relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
+	}
```
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:2337]: I think we want to wrap this inside a conditional `if shards > 1`

Seems wrong to add shard-related relabelings even though Prometheus isn't sharded 🤔

File: pkg/prometheus/promcfg.go (hunk 5 above)
Comment (ArthurSens) [reply pkg/prometheus/promcfg.go]: I think you removed the conditional you had above, which means we need the condition back 😅 

IMHO we should keep the conditional before we get here, i.e. on line 2326

File: pkg/prometheus/promcfg.go (hunk 7 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2337]: The initial reasoning was probably that the extra branch makes the code a bit less readable for little gain.

File: pkg/prometheus/promcfg.go [hunk 8]
```diff
@@ -2318,11 +2320,23 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
-	if sc.Spec.RelabelConfigs != nil {
+	if len(sc.Spec.RelabelConfigs) > 0 {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
-		cfg = append(cfg, yaml.MapItem{Key: "relabel_configs", Value: relabelings})
 	}
 
+	alreadyHasShardLabel := false
+	for _, relabelConfig := range sc.Spec.RelabelConfigs {
+		if relabelConfig.Action == "hashmod" {
+			level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "relabeling", relabelConfig)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2330]: (nit) RelabelConfig doesn't implement the Stringer interface so it won't log cleanly.
```suggestion
			level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "idx", i)
```

File: pkg/prometheus/promcfg.go [hunk 9]
```diff
@@ -2318,11 +2320,23 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
 	}
 
-	if sc.Spec.RelabelConfigs != nil {
+	if len(sc.Spec.RelabelConfigs) > 0 {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
-		cfg = append(cfg, yaml.MapItem{Key: "relabel_configs", Value: relabelings})
 	}
 
+	alreadyHasShardLabel := false
+	for _, relabelConfig := range sc.Spec.RelabelConfigs {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2328]:
```suggestion
	for i, relabelConfig := range sc.Spec.RelabelConfigs {
```

File: pkg/prometheus/promcfg.go (hunk 7 above)
Comment (ArthurSens) [reply pkg/prometheus/promcfg.go]: I imagine that Prometheus will need to perform extra CPU work for nothing if we add a no-op relabeling rule. I don't know exactly how much of an impact this could have, but I can imagine a few CPU cycles going to waste for very large-scale Prometheus :P

Comment (adinhodovic) [reply pkg/prometheus/promcfg.go]: I've reverted the change and the skipping non-sharded configs will be done again now. I agree on not adding a relabelling rule, might be confusing for users that do not shard as well.

File: pkg/prometheus/promcfg.go [hunk 10]
```diff
@@ -2318,8 +2320,25 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+				alreadyHasShardLabel = true
+				break
+			}
+		}
+		if !alreadyHasShardLabel {
+			relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
+		}
+	}
+
+	if len(sc.Spec.RelabelConfigs) > 0 || shards != 1 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2341]:
```suggestion
	if len(relabelings) > 0 {
```

File: pkg/prometheus/promcfg.go [hunk 11]
```diff
@@ -2318,8 +2320,25 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
 	}
 
-	if sc.Spec.RelabelConfigs != nil {
+	if len(sc.Spec.RelabelConfigs) > 0 {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if shards != 1 {
+		alreadyHasShardLabel := false
+		for i, relabelConfig := range sc.Spec.RelabelConfigs {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2329]: (nit)
```suggestion
		for i, relabelConfig := range relabelconfigs {
```

File: pkg/prometheus/promcfg.go [hunk 12]
```diff
@@ -2318,8 +2320,25 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			if relabelConfig.Action == "hashmod" {
+				level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "idx", i)
+				alreadyHasShardLabel = true
+				break
+			}
+		}
+		if !alreadyHasShardLabel {
+			relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2337]: (suggestion) we could also have a new `generateAddressShardingRelabelingRulesIfMissing` function (sorry for the long name, feel free to use something shorter!). It would make code a bit more readable IMHO.

**Review by ArthurSens — APPROVED** [review]
LGTM! Sorry for the whole ping-pong with the 1 shard situation 😬

File: pkg/prometheus/promcfg.go [hunk 13]
```diff
@@ -1444,6 +1444,22 @@ func generateAddressShardingRelabelingRules(relabelings []yaml.MapSlice, shards
...
+			if relabelItem.Key == "action" && relabelItem.Value == "hashmod" {
+				level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "idx", i)
+				alreadyHasShardLabel = true
+			}
+		}
+	}
+	if !alreadyHasShardLabel {
+		return generateAddressShardingRelabelingRules(relabelings, shards)
+	}
+	return relabelings
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1460]: just a suggestion, feel free to ignore
```suggestion
	for i, relabeling := range relabelings {
		for _, relabelItem := range relabeling {
			if relabelItem.Key == "action" && relabelItem.Value == "hashmod" {
				level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "idx", i)
				return generateAddressShardingRelabelingRules(relabelings, shards)
			}
		}
	}

	return relabelings
```

Comment (adinhodovic) [reply pkg/prometheus/promcfg.go]: cleaner :+1:

**Review by simonpasquier — APPROVED** [review]
Thanks!
